File: ml-worker/tests/fixtures.py
"""Synthetic dataset generators for the golden tests.

Three flavours, all returning a pandas DataFrame in the on-disk contract layout
(``ts``, ``t1``, ``label`` in R units, ``regime``, and ``layer__feature`` columns):

  * :func:`genuine_edge_dataset` — features genuinely (causally) predict the label, so a
    model trained on the past keeps an edge out-of-sample. Used to prove the harness does
    NOT reject good strategies (low PBO, positive DSR, gate PASSES).
  * :func:`pure_noise_dataset` — features are independent of the label. Any apparent edge
    is selection noise. Used to prove high PBO / failing gate.
  * :func:`leaky_dataset` — contains a future-peeking feature equal to (a noised copy of)
    the label itself. In-sample it looks spectacular; out-of-sample, once the leak is
    purged/embargoed and the model can't actually see the future, it collapses. THE
    leakage checkpoint.

Also provides :func:`make_perf_matrix_*` builders for the PBO test.
"""

from __future__ import annotations

import numpy as np
import pandas as pd

_REGIMES = np.array(["bull", "bear", "chop"])


def _timeline(n: int, start: str = "2020-01-01") -> tuple[pd.DatetimeIndex, pd.DatetimeIndex]:
    """Event timestamps and (event + hold) barrier timestamps on a daily grid."""
    ts = pd.date_range(start=start, periods=n, freq="D")
    # Each label window spans 3 bars (so adjacent events overlap — exercises purging).
    t1 = ts + pd.Timedelta(days=3)
    return ts, t1


def genuine_edge_dataset(n: int = 1500, seed: int = 0) -> pd.DataFrame:
    """Features causally drive the label; the edge is real and persistent."""
    rng = np.random.default_rng(seed)
    ts, t1 = _timeline(n)

    momentum = rng.normal(size=n)
    trend = rng.normal(size=n)
    vol = rng.normal(size=n)
    noise_feat = rng.normal(size=n)

    # A stable linear signal + noise drives the R outcome.
    signal = 0.9 * momentum + 0.6 * trend - 0.3 * vol
    label = signal + rng.normal(scale=1.0, size=n)
    # Express in R-ish units centred so there is a genuine positive expectancy.
    label = np.clip(label, -1.0, 2.0)

    regime = _REGIMES[rng.integers(0, 3, size=n)]
    return pd.DataFrame(
        {
            "ts": ts,
            "t1": t1,
            "label": label,
            "regime": regime,
            "momentum__signal": momentum,
            "trend__slope": trend,
            "volatility__atr_norm": vol,
            "momentum__noise": noise_feat,
        }
    )


def pure_noise_dataset(n: int = 1500, seed: int = 1) -> pd.DataFrame:
    """Features are independent of the label, and the label has ZERO expectancy.

    The label is symmetric around zero (no structural drift), so absent any real predictive
    feature the achievable edge is nil — and once a per-trade cost is subtracted, the
    cost-aware OOS expectancy is negative. The harness must therefore REJECT it.
    """
    rng = np.random.default_rng(seed)
    ts, t1 = _timeline(n)

    label = np.clip(rng.normal(scale=1.0, size=n), -1.5, 1.5)
    regime = _REGIMES[rng.integers(0, 3, size=n)]
    return pd.DataFrame(
        {
            "ts": ts,
            "t1": t1,
            "label": label,
            "regime": regime,
            "momentum__a": rng.normal(size=n),
            "trend__b": rng.normal(size=n),
            "volatility__c": rng.normal(size=n),
            "momentum__d": rng.normal(size=n),
        }
    )


def leaky_dataset(n: int = 1500, seed: int = 2) -> pd.DataFrame:
    """A planted look-ahead leak: spectacular in-sample, collapses out-of-sample.

    The leak feature reveals the label's sign — but only in the FIRST part of the
    timeline. In the held-out (later) part its sign FLIPS to point the wrong way::

        leak__lookahead[i] =  +label_signal[i]   for the early (in-sample) bars
                              -label_signal[i]   for the later (out-of-sample) bars

    This is the signature of look-ahead bias / a non-point-in-time feature: a quantity that
    *appears* to predict the label on the data the researcher fit on, but whose relationship
    does not hold on genuinely unseen, later data. Because the genuine features carry no
    edge, a naive search selects the leak-driven config on the in-sample slice (great IS
    metrics) and the validation harness — which selects IS but judges on the later OOS
    slice — sees it collapse: cost-aware OOS expectancy goes non-positive and DSR deflates,
    so the promotion gate REJECTS it. A genuine, stationary edge does not flip and passes.
    """
    rng = np.random.default_rng(seed)
    ts, t1 = _timeline(n)

    base = rng.normal(size=n)
    label = base + rng.normal(scale=1.0, size=n)
    label = np.clip(label, -1.5, 1.5)  # symmetric: no free drift once the leak flips

    # The leak tracks the label early, then its sign flips for the later (OOS) portion.
    sign = np.ones(n)
    sign[n // 2 :] = -1.0
    leak = sign * base + rng.normal(scale=0.01, size=n)

    regime = _REGIMES[rng.integers(0, 3, size=n)]
    return pd.DataFrame(
        {
            "ts": ts,
            "t1": t1,
            "label": label,
            "regime": regime,
            "leak__lookahead": leak,
            # Genuine features are pure noise: nothing real survives the regime flip.
            "momentum__a": rng.normal(size=n),
            "trend__b": rng.normal(size=n),
            "volatility__c": rng.normal(size=n),
        }
    )


# --------------------------------------------------------------------------- #
# PBO performance-matrix builders
# --------------------------------------------------------------------------- #
def perf_matrix_genuine(t: int = 800, n_strategies: int = 20, seed: int = 7) -> np.ndarray:
    """Performance matrix where ONE strategy has a persistent (IS+OOS) edge.

    The genuinely good strategy is good across the whole timeline, so CSCV's IS-best
    selection keeps choosing it and it stays above-median OOS -> low PBO.
    """
    rng = np.random.default_rng(seed)
    M = rng.normal(scale=1.0, size=(t, n_strategies))
    # Strategy 0 has a real, time-stable positive mean.
    M[:, 0] += 0.5
    return M


def perf_matrix_overfit(t: int = 800, n_strategies: int = 200, seed: int = 8) -> np.ndarray:
    """Many pure-noise strategies; the IS-best is best by luck -> high PBO.

    With many independent noise strategies, whichever wins in-sample is essentially random
    out-of-sample, so it lands below median OOS about half the time -> PBO ~ 0.5+.
    """
    rng = np.random.default_rng(seed)
    return rng.normal(scale=1.0, size=(t, n_strategies))

File: crates/se-validation/src/fixtures.rs
//! Deterministic datasets for the deliberate-leakage checkpoint (§8 / P4).
//!
//! Two generators with opposite truth:
//!   * [`genuine_edge_dataset`] — features causally drive the label across the whole
//!     timeline, so a purged+embargoed CPCV model keeps an edge out-of-sample.
//!   * [`leaky_dataset`] — a `leak__lookahead` feature equals the label's signal on the
//!     in-sample (early) half then FLIPS SIGN on the held-out (later) half. In-sample it
//!     looks spectacular; out-of-sample it collapses. The promotion gate MUST reject it.
//!
//! Shared by the integration test and the `inject-leak-test` CLI command so the operator
//! runs exactly what CI verifies.

use std::collections::BTreeMap;

use chrono::{DateTime, TimeZone, Utc};
use se_mlclient::DatasetRow;

/// Number of rows in each fixture.
pub const N: usize = 900;
/// Regime labels cycled through the fixtures (for the ≥2-regime gate condition).
pub const REGIMES: [&str; 3] = ["bull", "bear", "chop"];

/// A tiny deterministic PRNG (SplitMix64) — reproducible draws with the right
/// statistical structure (we do not need to reproduce numpy's RNG).
pub struct Rng(u64);

impl Rng {
    pub fn new(seed: u64) -> Self {
        Rng(seed.wrapping_add(0x9E37_79B9_7F4A_7C15))
    }

    pub fn next_u64(&mut self) -> u64 {
        self.0 = self.0.wrapping_add(0x9E37_79B9_7F4A_7C15);
        let mut z = self.0;
        z = (z ^ (z >> 30)).wrapping_mul(0xBF58_476D_1CE4_E5B9);
        z = (z ^ (z >> 27)).wrapping_mul(0x94D0_49BB_1331_11EB);
        z ^ (z >> 31)
    }

    /// Uniform in [0, 1) using a 53-bit mantissa.
    pub fn uniform(&mut self) -> f64 {
        (self.next_u64() >> 11) as f64 / (1u64 << 53) as f64
    }

    /// Standard normal via Box-Muller.
    pub fn normal(&mut self) -> f64 {
        let u1 = self.uniform().max(1e-12);
        let u2 = self.uniform();
        (-2.0 * u1.ln()).sqrt() * (std::f64::consts::TAU * u2).cos()
    }
}

fn ts(i: usize) -> DateTime<Utc> {
    // Daily grid; `t1` is +3 days so adjacent label windows overlap (exercises purging).
    Utc.with_ymd_and_hms(2020, 1, 1, 0, 0, 0).unwrap() + chrono::Duration::days(i as i64)
}

fn clip(x: f64, lo: f64, hi: f64) -> f64 {
    x.max(lo).min(hi)
}

/// Genuine edge: a stable linear signal of three features drives the R label across the
/// whole timeline.
pub fn genuine_edge_dataset(seed: u64) -> Vec<DatasetRow> {
    let mut rng = Rng::new(seed);
    let mut rows = Vec::with_capacity(N);
    for i in 0..N {
        let momentum = rng.normal();
        let trend = rng.normal();
        let vol = rng.normal();
        let noise_feat = rng.normal();

        let signal = 0.9 * momentum + 0.6 * trend - 0.3 * vol;
        let label = clip(signal + rng.normal(), -1.0, 2.0);

        let regime = REGIMES[(rng.next_u64() % 3) as usize];
        let features = BTreeMap::from([
            ("momentum__signal".to_string(), momentum),
            ("trend__slope".to_string(), trend),
            ("volatility__atr_norm".to_string(), vol),
            ("momentum__noise".to_string(), noise_feat),
        ]);
        rows.push(DatasetRow {
            ts: ts(i),
            t1: ts(i + 3),
            label,
            regime: Some(regime.to_string()),
            features,
        });
    }
    rows
}

/// Leaky look-ahead: `leak__lookahead` tracks the label's signal early, then FLIPS sign on
/// the later (held-out) half. The genuine features are pure noise, so a naive search locks
/// onto the leak in-sample and collapses out-of-sample.
pub fn leaky_dataset(seed: u64) -> Vec<DatasetRow> {
    let mut rng = Rng::new(seed);
    let mut rows = Vec::with_capacity(N);
    let half = N / 2;
    for i in 0..N {
        let base = rng.normal();
        let label = clip(base + rng.normal(), -1.5, 1.5); // symmetric: no free drift

        let sign = if i < half { 1.0 } else { -1.0 };
        let leak = sign * base + 0.01 * rng.normal();

        let a = rng.normal();
        let b = rng.normal();
        let c = rng.normal();

        let regime = REGIMES[(rng.next_u64() % 3) as usize];
        let features = BTreeMap::from([
            ("leak__lookahead".to_string(), leak),
            ("momentum__a".to_string(), a),
            ("trend__b".to_string(), b),
            ("volatility__c".to_string(), c),
        ]);
        rows.push(DatasetRow {
            ts: ts(i),
            t1: ts(i + 3),
            label,
            regime: Some(regime.to_string()),
            features,
        });
    }
    rows
}
